.cloudbuild/scripts/internal/changes/changes.go

```diff
@@ -0,0 +1,159 @@
+/*
+Copyright 2021 Gravitational, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// Package main implements a script to test if a PR being tested contains only
+// documentation changes. If so, it will create a signal file in a specified
+// location, the existence of which subsequent build steps can use to determine
+// if they should run or not.
+package changes
+
+import (
+	"log"
+	"os"
+	"strings"
+
+	"github.com/go-git/go-git/v5"
+	"github.com/go-git/go-git/v5/plumbing"
+	"github.com/go-git/go-git/v5/plumbing/object"
+	"github.com/gravitational/trace"
+)
+
+// Changes describes the kind of changes found in the analysed workspace.
+type Changes struct {
+	Docs bool
+	Code bool
+}
+
+// Analyze examines the workspace for specific changes using its git history,
+// and then collates and returns a report.
+func Analyze(workspaceDir string, targetBranch string, commitSHA string) (Changes, error) {
+	log.Printf("Opening workspace %q as git repo\n", workspaceDir)
+	repo, err := git.PlainOpen(workspaceDir)
+	if err != nil {
+		return Changes{}, trace.Wrap(err, "failed opening workspace as a repo")
+	}
+
+	changes, err := getChanges(repo, targetBranch, commitSHA)
+	if err != nil {
+		return Changes{}, trace.Wrap(err, "failed extracting changes")
+	}
+
+	report := Changes{}
+
+	for _, change := range changes {
+		if isDocChange(change) {
+			report.Docs = report.Docs || true
+		} else {
+			report.Code = report.Code || true
+		}
+
+		if report.Docs && report.Code {
+			// There's no sense in exhaustively listing all of the changes if
+			// the answer won't change, so bail early.
+			break
+		}
+	}
+
+	return report, nil
+}
+
+func isDocChange(change *object.Change) bool {
+	path := getChangePath(change)
+	return path == "" ||
+		strings.HasPrefix(path, "docs/") ||
+		strings.HasSuffix(path, ".mdx") ||
+		strings.HasSuffix(path, ".md")
```

Comment:
nit: in what case `path == ""`? and maybe also lower case `path` for comparison just to be safe.

Reply:
The `path == ""` case was a bad cargo-cult copy out of the original shell script - have removed it.